shell/platform/embedder/tests/embedder_unittests_metal.mm

```diff
@@ -227,6 +227,49 @@ GrBackendTexture backend_texture(texture_size.width(), texture_size.height(), Gr
   ASSERT_TRUE(ImageMatchesFixture("scene_without_custom_compositor.png", rendered_scene));
 }
 
+TEST_F(EmbedderTest, TextureDestructionCallbackCalledWithoutCustomCompositorMetal) {
+  EmbedderTestContextMetal& context = reinterpret_cast<EmbedderTestContextMetal&>(
+      GetEmbedderContext(EmbedderTestContextType::kMetalContext));
+  EmbedderConfigBuilder builder(context);
+  builder.SetMetalRendererConfig(SkISize::Make(800, 600));
+  builder.SetDartEntrypoint("texture_destruction_callback_called_without_custom_compositor");
+
+  struct CollectContext {
+    int collect_count = 0;
+    fml::AutoResetWaitableEvent latch;
+  };
+
+  auto collect_context = std::make_unique<CollectContext>();
+  context.SetNextDrawableCallback([&context, &collect_context](const FlutterFrameInfo* frame_info) {
+    auto texture_info = context.GetTextureInfo();
+    FlutterMetalTexture texture;
+    texture.struct_size = sizeof(FlutterMetalTexture);
+    texture.texture_id = texture_info.texture_id;
+    texture.texture = reinterpret_cast<FlutterMetalTextureHandle>(texture_info.texture);
+    texture.user_data = collect_context.get();
+    texture.destruction_callback = [](void* user_data) {
+      CollectContext* callback_collect_context = reinterpret_cast<CollectContext*>(user_data);
+      callback_collect_context->collect_count++;
+      callback_collect_context->latch.Signal();
```

Comment:
Latching/signalling here bit dirty since the test either passes or hangs until it times out and fails.

What I'm really trying to capture here is the moment where one complete frame has been drawn and the associated surface/texture collected.

Initially I tried to do a countdown latch on onBeginFrame. On the second call, we expect that the first surface has been disposed. This actually works pretty well except that I haven't found a nice mechanism to guarantee that we actually *get* the second frame before everything shuts down.

I wonder if I could use a Completer on the dart side in onDraw in the dart code and schedule the frame, wait on it, then schedule the next frame and wait on that. Will give it a shot and report back here. Any ideas appreciated though :)

Reply:
I believe we have other spots where this pattern if followed. The hanging case is handled by the per test case watchdog which only allows a test to run for a small amount of time.